kernel/kernel_drivers.go

```diff
@@ -179,34 +202,129 @@ func RemoveKernelDriversTree(ksnapName string, rev snap.Revision) (err error) {
+// currently active kernel.
+func EnsureKernelDriversTree(ksnapName string, rev snap.Revision, kernelMount string, kmodsInfos []*KernelModulesMntInfo, opts *KernelDriversTreeOptions) (err error) {
+	// The temporal dir when installing only components can be fixed as a
+	// task installing/updating a kernel-modules component must conflict
+	// with changes containing this same task. This helps with clean-ups if
+	// something goes wrong. Note that this folder needs to be in the same
+	// filesystem as the final one so we can atomically switch the folders.
+	ksnapDir := ksnapName + "_tmp"
+	if opts.KernelInstall {
+		ksnapDir = ksnapName
+	}
 	// Initial clean-up to make the function idempotent
-	if rmErr := RemoveKernelDriversTree(ksnapName, rev); rmErr != nil &&
+	if rmErr := RemoveKernelDriversTree(ksnapDir, rev); rmErr != nil &&
 		!errors.Is(err, fs.ErrNotExist) {
 		logger.Noticef("while removing old kernel tree: %v", rmErr)
 	}
 
-	// Remove tree in case something goes wrong
 	defer func() {
-		if err != nil {
-			if rmErr := RemoveKernelDriversTree(ksnapName, rev); rmErr != nil &&
-				!errors.Is(err, fs.ErrNotExist) {
-				logger.Noticef("while cleaning up kernel tree: %v", rmErr)
-			}
+		// Remove on return if error or if temporal tree
+		if err == nil && opts.KernelInstall {
+			return
+		}
+		if rmErr := RemoveKernelDriversTree(ksnapDir, rev); rmErr != nil &&
+			!errors.Is(err, fs.ErrNotExist) {
+			logger.Noticef("while cleaning up kernel tree: %v", rmErr)
 		}
 	}()
 
-	// Root of our kernel tree
-	treeRoot := driversTreeDir(ksnapName, rev)
+	treeRoot := driversTreeDir(ksnapDir, rev)
 
-	if err := createModulesSubtree(kernelMount, treeRoot); err != nil {
+	// Create drivers tree
+	kversion, err := KernelVersionFromModulesDir(kernelMount)
+	if err == nil {
+		if err := createModulesSubtree(kernelMount, treeRoot,
+			kversion, kmodsInfos); err != nil {
+			return err
+		}
+	} else {
+		// Bit of a corner case, but maybe possible. Log anyway.
+		// TODO detect this issue in snap pack, should be enforced
+		// if the snap declares kernel-modules components.
+		logger.Noticef("no modules found in %q", kernelMount)
+	}
+
+	fwDir := filepath.Join(treeRoot, "lib", "firmware")
+	if opts.KernelInstall {
+		// symlinks in /lib/firmware are not affected by components
+		if err := createFirmwareSymlinks(kernelMount, fwDir); err != nil {
+			return err
+		}
+	}
+	updateFwDir := filepath.Join(fwDir, "updates")
+	// This folder needs to exist always to allow for directory swapping
+	// in the future, even if right now we don't have components.
+	if err := os.MkdirAll(updateFwDir, 0755); err != nil {
 		return err
 	}
+	for _, kmi := range kmodsInfos {
+		if err := createFirmwareSymlinks(kmi.MntDir, updateFwDir); err != nil {
+			return err
+		}
+	}
+
+	// Sync before returning successfully (install kernel case) and also
+	// for swapping case so we have consistent content before swapping
+	// folder.
+	syscall.Sync()
```

Comment:
I completely forgot about it, but we have something called `osutil.AtomicSymlink`, which does sync whenever needed under the hood. Perhaps the `create*` helpers could be updated.

Reply:
Here we are creating quite a few symlinks (~500) in the firmware folder, and for modules we copy files and call depmod that writes files as well, so probably syncing here is cheaper than doing it per symlink/files (and we do not have control on depmod writes).

Reply:
makes sense, thanks